CodeTaskFactory can cause later task failures by loading a redundant copy of Microsoft.Build.Framework

This was reported and investigated (thanks!) offline by @KevinRansom.

The F# build failed when built with 64-bit MSBuild, with a failure like this:

```
error MSB4018: The "GetPackOutputItemsTask" task failed unexpectedly. [S:\fsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
error MSB4018: System.ArrayTypeMismatchException: Attempted to access an element as a type incompatible with the array. [S:\fsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
error MSB4018:    at System.Collections.Generic.List`1.Add(T item) [S:\fsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
error MSB4018:    at NuGet.Build.Tasks.Pack.GetPackOutputItemsTask.Execute() [S:\fsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [S:\fsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [S:\fsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
```

Debugging revealed that multiple copies of `Microsoft.Build.Framework.dll` were loaded--one from the `MSBuild\bin` directory, and one from `MSBuild\bin\amd64`.

The former makes sense because of the codebase we specify

https://github.com/dotnet/msbuild/blob/01e6196d5df534dec8c16a1417ff04638330c790/src/MSBuild/app.amd64.config#L19-L23

But the latter did not.

After some more debugging we found that the assembly was being loaded here

https://github.com/dotnet/msbuild/blob/01e6196d5df534dec8c16a1417ff04638330c790/src/Tasks/CodeTaskFactory.cs#L680-L686

As a result of entries in a `CodeTaskFactory` task definition:

```xml
      <Reference Include="$(MSBuildToolsPath)\Microsoft.Build.Framework.dll"/>
      <Reference Include="$(MSBuildToolsPath)\Microsoft.Build.Utilities.Core.dll"/>
```

In 64-bit MSBuild, `$(MSBuildToolsPath)` resolves to the `amd64` folder, passing that copy of `Microsoft.Build.Framework.dll` to `UnsafeLoadFrom`.

We then ignore the reference because it's provided by default

https://github.com/dotnet/msbuild/blob/01e6196d5df534dec8c16a1417ff04638330c790/src/Tasks/CodeTaskFactory.cs#L689-L696

But having already loaded it, it becomes possible to get incompatible definitions of `ITaskItem`, which is what happened to the ([perfectly reasonably authored](https://github.com/NuGet/NuGet.Client/blob/e0edb52d2ee204ab1117c9a592addc705cc76471/src/NuGet.Core/NuGet.Build.Tasks.Pack/GetPackOutputItemsTask.cs#L57-L59)) NuGet task.

```c#
            var outputs = new List<ITaskItem>();
            outputs.Add(new TaskItem(Path.Combine(PackageOutputPath, nupkgFileName)));
            outputs.Add(new TaskItem(Path.Combine(NuspecOutputPath, nuspecFileName)));
```

--- a/src/Tasks/CodeTaskFactory.cs
+++ b/src/Tasks/CodeTaskFactory.cs
@@ -683,18 +683,24 @@ namespace Microsoft.Build.Tasks
 
                 try
                 {
+                    // Framework and Utilities are default references but are often
+                    // specified in the UsingTask anyway; if so just ignore them.
+                    //
+                    // Do this with an explicit upfront check rather than loading the
+                    // assembly and then checking its name, because that can cause
+                    // the loader to have multiple copies of these assemblies as in
+                    // https://github.com/dotnet/msbuild/issues/7108.
+
+                    string name = AssemblyName.GetAssemblyName(assemblyFile).FullName;
+                    if (name == _msbuildFrameworkName ||
+                        name == _msbuildUtilitiesName)
+                    {
+                        return false;
+                    }
+
                     Assembly candidateAssembly = Assembly.UnsafeLoadFrom(assemblyFile);
                     if (candidateAssembly != null)
                     {
-                        string name = candidateAssembly.FullName;
-                        if (name == _msbuildFrameworkName ||
-                            name == _msbuildUtilitiesName)
-                        {
-                            // Framework and Utilities are default references but are often
-                            // specified in the UsingTask anyway; if so just ignore them.
-                            return false;
-                        }
-
                         candidateAssemblyLocation = candidateAssembly.Location;
                         s_knownReferenceAssemblies[candidateAssembly.FullName] = candidateAssembly;
                     }

